pkg/src/geolocation/GeolocationBase.ts

```diff
@@ -0,0 +1,78 @@
+import axios from 'axios'
+
+/**
+ * Geo location abstraction
+ * having an address an endpoint and an this.apiKey,
+ * this class is able to calculate latitude and longitude coordinates with a 3rd party service
+ */
+export default abstract class GeolocationBase {
+	lat?: number
+
+	lon?: number
+
+	apiKey? = ``
+	endpoint?=``
+
+	isPublic = false
+
+	address: string
+
+	/**
+	 * default constructor
+	 * @param address the addres to have it's coordinates fetched
+	 */
+	constructor(address: string, isPublic: boolean, apiKey: string, endpoint: string) {
+		this.apiKey = apiKey
+		this.endpoint = endpoint
+
+		this.address = address
+		this.isPublic = isPublic
+	}
+
+	/**
+	 * creates a custom request's url
+	 * IF overriden, it's outpur will override the createUrl's
+	 */
+	createCustomUrl(): string | null {
+		return null
+	}
+
+	/**
+	 * parses the get geocode response, into valid latitude and longitude values
+	 * @returns the pair of latitude and longitude values (if contained in the given payload)
+	 */
+	// eslint-disable-next-line
+	abstract parseResult(payload: {}): { lat: number, lon: number }
+
+	/**
+	 * creates the specific request url for this vendor
+	 */
+	createUrl(): string {
+		if (!this.apiKey || this.apiKey === `process.env.MAP_BOX_ACCESS_KEY` || this.apiKey === `MAP_BOX_ACCESS_KEY`)
+			throw new Error(`in order to fetch a geolocation we need an Provider api key`)
+
+		if (!this.endpoint) throw Error(`impossible to create a request url without the vendor endpoint`)
+		if (!this.isPublic) {
+			if (!this.apiKey) throw Error(`impossible to create a request url without the vendor this.apiKey`)
+
+			// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+			return this.endpoint.replace(`@`, escape(this.address)).replace(`#`, this.apiKey)
+		}
+		return this.endpoint.replace(`@`, escape(this.address))
+	}
+
+	/**
+	 * actually fetches a set of coordinates for the given address
+	 */
+	async getCoordinates(): Promise<void> {
+		const url = this.createUrl()
+		const response = await axios.get(url)
```

Comment:
How would you refactor this base service if we had to support some other provider that required us to pass credentials via the header or make calls via POST?

Reply:
I'd make the GeolocationBase implement all three ways of making requests, and have it receive a configuration object, specifying the request method.

With the current request method in hands, the instance would be able to fetch data accordingly

Reply:
Do you see any drawbacks of adding a configuration object + cyclomatic complexity to handle this scenario?